controller/src/lib.rs

```diff
@@ -508,68 +512,173 @@ impl Controller {
+            })
     }
 
     /// Reset a set of transceiver modules.
     pub async fn reset(&self, _modules: ModuleId) -> Result<(), Error> {
         todo!()
     }
 
-    /// Set the power mode for a set of transceiver modules.
-    pub async fn set_power_mode(&self, _modules: ModuleId, _mode: PowerMode) -> Result<(), Error> {
-        todo!()
+    // Fetch the software power control state of a set of modules.
+    async fn power_control(&self, modules: ModuleId) -> Result<Vec<PowerControl>, Error> {
+        self.parse_modules_by_identifier::<PowerControl>(modules)
+            .await
+            .map(|collection| {
+                collection
+                    .into_values()
+                    .map(|(_id, control)| control)
+                    .collect()
+            })
+    }
+
+    // Return the subset of `modules` where `f(status) == true`.
+    fn filter_modules_with<F>(modules: ModuleId, status: &[Status], f: F) -> Result<ModuleId, Error>
+    where
+        F: Fn(Status) -> bool,
+    {
+        PortMask::from_index_iter(
+            modules
+                .ports
+                .to_indices()
+                .zip(status.into_iter())
+                .filter_map(|(ix, st)| if f(*st) { Some(ix) } else { None }),
+        )
+        .map(|ports| ModuleId {
+            fpga_id: modules.fpga_id,
+            ports,
+        })
+        .map_err(Error::from)
+    }
+
+    /// Get the power mode of a set of transceiver modules.
+    ///
+    /// For each module, this returns the actual `PowerMode`, as well as whether
+    /// the module has set software-override of power control. In the case where
+    /// the module is in off, that can't be determined, and `None` is returned.
+    pub async fn power_mode(
+        &self,
+        modules: ModuleId,
+    ) -> Result<Vec<(PowerMode, Option<bool>)>, Error> {
+        // Split the requested modules into those with power enabled via the
+        // e-fuse, and those without. The latter are always reported as off.
+        let status = self.status(modules).await?;
+        let unpowered_modules = Self::filter_modules_with(modules, &status, |status| {
+            !status.contains(Status::POWER_GOOD | Status::ENABLED)
+        })?;
+        let powered_modules = ModuleId {
+            fpga_id: modules.fpga_id,
+            ports: modules.ports.remove(unpowered_modules.ports),
+        };
+
+        // Of the powered modules, those in reset must be also be considered
+        // Off.
+        //
+        // Filter down the status of all modules to those that are powered.
+        let powered_status: Vec<_> = status
+            .iter()
+            .copied()
+            .enumerate()
+            .filter_map(|(ix, st)| {
+                let index = u8::try_from(ix).expect("Impossible index");
+                if powered_modules.contains(index) {
+                    Some(st)
+                } else {
+                    None
+                }
+            })
```

Comment:
Personal opinion: I'm not sold on `filter_map` being simpler than splitting up the filter and map, since you have to build the `Option` yourself...

(this also applies below)

```suggestion
            .filter(|(ix, _)| {
                let index = u8::try_from(ix).expect("Impossible index");
                powered_modules.contains(index)
            })
            .map(|(_, st)| st)
```